File: packages/nim/oauth/src/oauth_do/storage.nim
## File-based token storage.

import std/[json, os]
import device

let
  TokenDir = getHomeDir() / ".oauth.do"
  TokenFile = TokenDir / "token"

type
  TokenStorage* = object
    tokenDir: string
    tokenFile: string

proc newTokenStorage*(tokenDir: string = TokenDir): TokenStorage =
  ## Creates a new token storage.
  TokenStorage(
    tokenDir: tokenDir,
    tokenFile: tokenDir / "token"
  )

proc saveTokens*(storage: TokenStorage, tokens: TokenResponse) =
  ## Saves tokens to file storage.
  createDir(storage.tokenDir)
  writeFile(storage.tokenFile, $tokens.toJson())
  # Note: Nim doesn't have a built-in chmod, permissions depend on umask

proc loadTokens*(storage: TokenStorage): TokenResponse =
  ## Loads tokens from file storage.
  if not fileExists(storage.tokenFile):
    raise newException(IOError, "Token file not found")

  let content = readFile(storage.tokenFile)
  let json = parseJson(content)

  TokenResponse(
    accessToken: json["access_token"].getStr(),
    refreshToken: json{"refresh_token"}.getStr(""),
    tokenType: json{"token_type"}.getStr("Bearer"),
    expiresIn: json{"expires_in"}.getInt(0)
  )

proc getAccessToken*(storage: TokenStorage): string =
  ## Gets the access token from storage.
  try:
    let tokens = storage.loadTokens()
    tokens.accessToken
  except:
    ""

proc getRefreshToken*(storage: TokenStorage): string =
  ## Gets the refresh token from storage.
  try:
    let tokens = storage.loadTokens()
    tokens.refreshToken
  except:
    ""

proc deleteTokens*(storage: TokenStorage) =
  ## Deletes stored tokens.
  if fileExists(storage.tokenFile):
    removeFile(storage.tokenFile)

proc hasTokens*(storage: TokenStorage): bool =
  ## Checks if tokens exist.
  fileExists(storage.tokenFile)

File: packages/elixir/oauth/lib/oauth_do/storage.ex
defmodule OauthDo.Storage do
  @moduledoc """
  File-based token storage.
  """

  @token_dir Path.expand("~/.oauth.do")
  @token_file Path.join(@token_dir, "token")

  @doc """
  Saves tokens to file storage.
  """
  def save_tokens(tokens) do
    File.mkdir_p!(@token_dir)
    File.write!(@token_file, Jason.encode!(tokens))
    :ok
  end

  @doc """
  Loads tokens from file storage.
  """
  def load_tokens do
    case File.read(@token_file) do
      {:ok, content} -> {:ok, Jason.decode!(content)}
      {:error, :enoent} -> {:error, :not_found}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Gets the access token from storage.
  """
  def get_access_token do
    case load_tokens() do
      {:ok, %{"access_token" => token}} -> token
      _ -> nil
    end
  end

  @doc """
  Gets the refresh token from storage.
  """
  def get_refresh_token do
    case load_tokens() do
      {:ok, %{"refresh_token" => token}} -> token
      _ -> nil
    end
  end

  @doc """
  Deletes stored tokens.
  """
  def delete_tokens do
    case File.rm(@token_file) do
      :ok -> :ok
      {:error, :enoent} -> :ok
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Checks if tokens exist.
  """
  def has_tokens? do
    File.exists?(@token_file)
  end
end
